TagsCloudVisualization Task

Andreyinthesky opened the pull request.

Commit:
Init TagsCloudVisualization project

Commit:
Complete TagsCloudVisualization task
Added some test cases, tests moved to .cs file
Added visualization image

Commit:
Change encoding in README.md

Griboedoff commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,72 @@
+    {
+        public Point Center { get; }
+
+        private List<Rectangle> layoutedRectangles;
```
🤔 интересное слово `layouted`

Griboedoff commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,72 @@
+            return rectangle;
+        }
+
+        private bool RectangleIsIntersectOthers(Rectangle rectangle)
```
Название понятно, но сказано не по-английски

Griboedoff commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+            var rectangle = new Rectangle();
+
+            foreach (var spiralPoint in spiral.GetSpiralPoints())
```
каждый раз начинаешь итерироваться по спирали с центра
давай сохранять положение

Griboedoff reviewed (changes requested):
нужно больше тестов
+ перфоманс тесты
+ тесты на плотность

Commit:
Fixes
Add Point, Rectangle Extensions, + TESTS
Add Layouter perfomance and tight TESTS
Add CloudVisualizer
+ more text format fixes

Commit:
Improved Layouter performance

Commit:
Update Readme.md
+ improved Visualizer

Griboedoff commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,140 @@
+            return ShiftRectangleToCloudCenterAlongY(ShiftRectangleToCloudCenterAlongX(rect));
+        }
+
+        private Rectangle ShiftRectangleToCloudCenterAlongX(Rectangle rect)
```
очень похож на ShiftRectangleToCloudCenterAlongY
давай попробуем обобщить

Griboedoff commented on cs/TagsCloudVisualization/CircularCloudLayouter_Should.cs:
```diff
@@ -0,0 +1,195 @@
+            {
+                var directory = TestContext.CurrentContext.TestDirectory;
+                var testName = TestContext.CurrentContext.Test.Name;
+                var path = $"{directory}/{testName}.png";
```
path combine тебе в помошь

Griboedoff reviewed (commented):
Вроде норм
Будет хорошо, если вынесешь тесты в отдельную папку

Commit:
Fixes
Extensions moved to own folder
Tests moved to own folder
Update Layouter
Update Layouter tests
